jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/dialects/bigquery/BigQueryQueryRewriter.java

```diff
@@ -0,0 +1,161 @@
+package com.exasol.adapter.dialects.bigquery;
+
+import java.sql.*;
+import java.util.logging.*;
+import java.util.regex.*;
+
+import com.exasol.*;
+import com.exasol.adapter.*;
+import com.exasol.adapter.dialects.*;
+import com.exasol.adapter.jdbc.*;
+import com.exasol.adapter.sql.*;
+
+/**
+ * This class implements a BigQuery-specific query rewriter
+ */
+public class BigQueryQueryRewriter extends BaseQueryRewriter {
+    private static final Logger LOGGER = Logger.getLogger(BigQueryQueryRewriter.class.getName());
+    private static final double[] TEN_POWERS = { 10d, 100d, 1000d, 10000d, 100000d, 1000000d };
+    private static final Pattern DATE_PATTERN = Pattern.compile("(\\d{4})-(\\d{1,2})-(\\d{1,2})");
+    private static final Pattern TIME_PATTERN = Pattern.compile("(\\d{1,2}):(\\d{1,2}):(\\d{1,2})(?:\\.(\\d{1,6})?)?");
+
+    /**
+     * Create a new instance of a {@link BigQueryQueryRewriter}
+     *
+     * @param dialect              dialect
+     * @param remoteMetadataReader remote metadata reader
+     * @param connection           JDBC connection to remote data source
+     */
+    public BigQueryQueryRewriter(final SqlDialect dialect, final RemoteMetadataReader remoteMetadataReader,
+            final Connection connection) {
+        super(dialect, remoteMetadataReader, connection);
+    }
+
+    @Override
+    public String rewrite(final SqlStatement statement, final ExaMetadata exaMetadata,
+            final AdapterProperties properties) throws AdapterException, SQLException {
+        final StringBuilder builder = new StringBuilder();
+        final String query = getQueryFromStatement(statement, properties);
+        LOGGER.fine(() -> "Query to rewrite: " + query);
+        try (final ResultSet resultSet = this.connection.createStatement().executeQuery(query)) {
+            builder.append("SELECT * FROM VALUES");
+            int rowNumber = 0;
+            while (resultSet.next()) {
+                if (rowNumber > 0) {
+                    builder.append(",");
+                }
+                appendRow(builder, resultSet, resultSet.getMetaData());
```

Comment:
Store metadata in a variable before the while loop and use this here.
This is theoretically better from a performance perspective in a multi-row scenario, even if in Java the compiler might optimize the consecutive access via method.

Reply:
done